Add files needed for app-sre pipeline

dak1n1 opened the pull request:
Added the generic scripts, along with a customized version of
olm-artifacts-template.yaml which includes all of the manifests
from the operator's deploy/ dir (plus namespace). This file contains
the input parameters used by the CI pipeline to deploy the operator via OLM.

https://jira.coreos.com/browse/SREP-2049

dak1n1 commented:
@jewzaam thanks for that info. I'm attempting to follow some documentation to get managed-velero-operator into the app-sre pipeline, but it doesn't mention how to do that for operators that don't live in Hive. It said specifically to copy the `./hack` directory from another operator. So that's what this PR is. But it sounds like I need the equivalent of that hack directory, but tailored to non-hive operators.

Is that handled in the makefile instead? Like here? https://github.com/openshift/dedicated-admin-operator/blob/master/project.mk

jharrington22 commented on hack/olm-registry/olm-artifacts-template.yaml:
```diff
@@ -0,0 +1,41 @@
+    channel: ${CHANNEL}
+    name: managed-velero-operator
+    source: managed-velero-operator-catalog
+    sourceNamespace: managed-velero-operator
```
@dak1n1 This needs to be a selector sync set containing the OLM object :)  like https://github.com/openshift/configure-alertmanager-operator/blob/master/build/templates/olm-artifacts-template.yaml.tmpl

jharrington22 replied:
You'll also need to create the namespace, role, robebinding and service account, also specify the namespaces in the OLM spec for the CatalogSource, OperatorGroup and Subscription see above dedicated-admin repo

jharrington22 commented on hack/generate-operator-bundle.py:
```diff
@@ -0,0 +1,87 @@
+if not os.path.exists(version_dir):
+    os.mkdir(version_dir)
+
+with open('config/templates/{}-csv-template.yaml'.format(OPERATOR_NAME), 'r') as stream:
```
You'll need to include this CSV template, use an existing operators template for reference

dak1n1 commented on config/templates/managed-velero-operator-csv-template.yaml:
```diff
@@ -0,0 +1,50 @@
+kind: ClusterServiceVersion
+metadata:
+  name: managed-velero-operator-0.0.1
+  namespace: openshift-velero
```
I'm not sure if this line is ok, since the other operators have `namespace: placeholder` here. But I wanted to fill it in with a non-placeholder value if possible.

jharrington22 replied:
The reason value is set to `placeholder` is because this CSV is not directly applied to the cluster, its installed by OLM in the correct namespace when the Catalog operator is notified of a new version through a subscription which is watching the operators catalog source - built and deployed by the app-sre pipleine here (see the `hack/app_sre_create_image_catalog.sh` script).

I'd prefer to keep this as `placeholder` as it matches our other operators and follows the standards in the document here (See CSV Metadata): https://github.com/operator-framework/operator-lifecycle-manager/blob/master/doc/design/building-your-csv.md

c-e-brumm commented:
I like it, but I can't find the changes requested from the other reviews. Other wise, LGTM

dak1n1 commented:
I believe these are the changes that were requested:

Adding the CSV template: https://github.com/openshift/managed-velero-operator/blob/2b32579862f75eb51637a59a9e673b504dc9cfc1/config/templates/managed-velero-operator-csv-template.yaml

And creating the SelectorSyncSet: https://github.com/openshift/managed-velero-operator/blob/2b32579862f75eb51637a59a9e673b504dc9cfc1/hack/olm-registry/olm-artifacts-template.yaml#L19

jharrington22 commented on hack/olm-registry/olm-artifacts-template.yaml:
```diff
@@ -0,0 +1,327 @@
+parameters:
+- name: REGISTRY_IMG
+  required: true
+- name: CHANNEL
```
Could we add `value: staging` here for the default

jharrington22 commented:
Hey @dak1n1 could you incorporate the changes made here https://github.com/openshift/aws-account-operator/pull/211 to your `hack/generate-operator-bundle.py` script?

dak1n1 commented:
@jharrington22 I added the changes from openshift/aws-account-operator#211 to `hack/generate-operator-bundle.py`, but I deleted the block about `deploy/uhc_cluster_role.yaml`, since that only seemed relevant to the aws-account-operator.

jharrington22 commented on hack/generate-operator-bundle.py:
```diff
@@ -0,0 +1,117 @@
+    # Update CSV template customresourcedefinitions key
+    csv['spec']['customresourcedefinitions']['owned'].append(
+        {
+            "description": "",
```
So my fault here but OLM doesn't like a blank description so I used the Kind name here as well as I missed the `name` key. See PRs #213 and #214 in the aws-account-operator repo.

dak1n1 replied:
Got it. I rebased today and added this to hack/generate-operator-bundle.py:

```
            "name": crd["metadata"]["name"],
            "description": crd["spec"]["names"]["kind"],
```

jharrington22 commented on hack/generate-operator-bundle.py:
```diff
@@ -0,0 +1,118 @@
+            "version": crd["spec"]["version"]
+        }
+    )
+
```
Totally a nit pick but 1 blank line here

jharrington22 commented on hack/generate-operator-bundle.py:
```diff
@@ -0,0 +1,118 @@
+
+
+
+# Copy all prometheus yaml files over to the bundle output dir:
```
This directory doesn't exist so you can remove all this.

cblecker commented on config/templates/managed-velero-operator-csv-template.yaml:
```diff
@@ -0,0 +1,50 @@
+    description: Operator to manage installation of Velero in OpenShift.
+    containerImage: quay.io/app-sre/managed-velero-operator:latest
+    createdAt: "2019-11-11T11:00:00Z"
+    support: Stefanie Forrester
```
Do we put individuals names in here?

dak1n1 replied:
Yes, I copied this from other operators. Though I committed this is back when you were on a different team, so now that you're back I'd be happy to put your name there instead :)

```
[dakini@nibbana openshift]$ find -name "*-csv-template.yaml" -exec grep "support: " '{}' \;
    support: Sam Nguyen
    support: Devan Goodwin
    support: Matt Woodson
    support: Carl Brumm
    support: Stefanie Forrester
    support: Wesley Hearn
    support: James Harrington
[dakini@nibbana openshift]$ find -name "*-csv-template.yaml"
./rbac-permissions-operator/config/templates/rbac-permissions-operator-csv-template.yaml
./hive/config/templates/hive-csv-template.yaml
./deadmanssnitch-operator/config/templates/deadmanssnitch-operator-csv-template.yaml
./certman-operator/config/templates/certman-operator-csv-template.yaml
./managed-velero-operator/config/templates/managed-velero-operator-csv-template.yaml
./pagerduty-operator/config/templates/pagerduty-operator-csv-template.yaml
./aws-account-operator/config/templates/aws-account-operator-csv-template.yaml
```

cblecker commented on config/templates/managed-velero-operator-csv-template.yaml:
```diff
@@ -0,0 +1,50 @@
+  - type: OwnNamespace
+    supported: true
+  - type: SingleNamespace
+    supported: true
```
I think this should be false.. the OperatorGroup should exist in `openshift-velero`

dak1n1 replied:
I could be wrong, but I imagine this setting is correct because the other operators seem to have this same setting:

```
  installModes:
  - type: OwnNamespace
    supported: true
  - type: SingleNamespace
    supported: true
  - type: MultiNamespace
    supported: false
  - type: AllNamespaces
    supported: false
```

I think it just means that the operator is scoped to a namespace instead of being scoped cluster-wide. Let me know if I'm missing something. I didn't really understand how the `OperatorGroup` affects this. But I will be working with James on Monday to deploy this, so I'll run it by him too.

cblecker commented on config/templates/managed-velero-operator-csv-template.yaml:
```diff
@@ -0,0 +1,50 @@
+  install:
+    strategy: deployment
+    spec:
+      clusterPermissions:
```
There should be a `permissions` section too, as this operator has both cluster scoped permissions, and NS scoped permissions

dak1n1 replied:
ok, I ran this command to grab a list of the permissions and place them into the csv template.

```
operator-sdk olm-catalog gen-csv --csv-version 0.0.1
```

cblecker commented on hack/olm-registry/olm-artifacts-template.yaml:
```diff
@@ -0,0 +1,328 @@
+            serviceAccountName: managed-velero-operator
+            containers:
+              - name: managed-velero-operator
+                image: quay.io/openshift-sre/managed-velero-operator
```
Shouldn't this reference the image tag?

cblecker commented:
/test all

cblecker commented on hack/olm-registry/olm-artifacts-template.yaml:
```diff
@@ -0,0 +1,328 @@
+            - s3:PutLifecycleConfiguration
+            resource: "*"
+    - apiVersion: apps/v1
+      kind: Deployment
```
Actually, is this deployment required at all? Isn't the deployment created by OLM?

jharrington22 commented on config/templates/managed-velero-operator-csv-template.yaml:
```diff
@@ -0,0 +1,125 @@
+    - kind: Velero
+      name: veleros.managed.openshift.io
+      version: v1alpha1
+  description: Operator to manage installation of Velero in OpenShift.
```
I just noticed, this indentation is incorrect see: https://gitlab.cee.redhat.com/service/saas-aws-account-operator-bundle/blob/staging/aws-account-operator/0.1.438-96d664f/aws-account-operator.v0.1.438-96d664f.clusterserviceversion.yaml#L16-20

jharrington22 replied:
I just noticed, this indentation is incorrect see: https://gitlab.cee.redhat.com/service/saas-aws-account-operator-bundle/blob/staging/aws-account-operator/0.1.438-96d664f/aws-account-operator.v0.1.438-96d664f.clusterserviceversion.yaml#L16-20

cblecker commented:
/hold
holding this for now
